sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java

```diff
@@ -270,20 +276,19 @@ public void validate() {
       }
       codec = metadata.getCodec();
       syncMarker = metadata.getSyncMarker();
-      fileSchemaString = metadata.getSchemaString();
-      // If the source was created with a null schema, use the schema that we read from the file's
-      // metadata.
-      if (readSchemaString == null) {
-        readSchemaString = metadata.getSchemaString();
+
+      // Only store both the read and file schema if they differ
```

Comment:
As long as we store the same reference java serialization will only encode the object once and store two object ids here in place instead.

Will update this to just always intern the file schema which should become the read schema string reference if they are equivalent.